docs/enhancements/bmac-reboot-hosts-on-iso-change.md

```diff
@@ -0,0 +1,132 @@
+last-updated: 2021-07-01
+---
+
+# BMAC reboots only those hosts impacted by an ISO change
+
+## Summary
+
+When an InfraEnv's ISO gets regenerated based on a change such as adding or
+changing an NMStateConfig resource, there may be only a subset of that
+InfraEnv's Agents that are affected by the change. Rather than reboot all of
+them with the newly-generated ISO, only those hosts that are affected by the
+new or modified NMStateConfig should be rebooted.
+
+## Motivation
+
+If there are 500 NMStateConfigs and 500 Agents all related to the same
+InfraEnv, and then someone adds the 501st NMStateConfig, we do not want to
+reboot all 500 Agents. Rebooting is an expensive and time-consuming operation.
+We need a way to determine which Agents or hosts should be rebooted in response
+to an NMStateConfig.
+
+### Goals
+
+* Avoid rebooting un-provisioned BareMetalHosts when it is not needed.
+* Leave the door open for other platforms (include None/boot-it-yourself) to also avoid host reboots that are not needed.
+
+### Non-Goals
+
+* 
+
+## Proposal
+
+The assisted-service will track whether a particular Agent was booted using an
+old InfraEnv or NMStateConfig. It will ensure that any Agent that was booted
+using obsolete artifacts will have:
+* a label indicating that it was booted from obsolete artifacts
+* an array in its Status of references to the resources that are obsolete
+
+### assisted-service InfraEnv controller
+
+The InfraEnv ISO URL will change when the ISO contents change. This will make
+it obvious to an API user that they have booted a host from an obsolete ISO,
+because the URLs won't match. A side benefit of ensuring that the ISO URL
+uniquely identifies the resource's contents, is that pull-through caches will
+have an easy time understanding when an file is no longer relevant.
+
+When assisted-service creates an ISO, it will capture and include in the ISO:
+* the name, namespace, UID, and generation of the InfraEnv
+* the name, namespace, UID, and generation of each NMStateConfig
+
+At runtime, those resource identifiers for the InfraEnv and the NMStateConfig
+it used will be passed by the agent to assisted-service and compared against
+the current resource generations to determine if they are obsolete.
+
+### assisted-service Agent controller
+
+When the Agent resource gets created, its status should include the
+above-described reference details for:
+* The InfraEnv
+* The NMStateConfig it utilized during its boot
+
+If those details indicate that any of the referenced resources have changed,
+then the controller will add a label to the Agent resource indicating such. It
+will also add a reference to each resource to an array in the Agent's Status.
+
+### Bare Metal Agent Controller
+
+The Bare Metal Agent Controller will watch for Agents to have that label and
+reboot the corresponding BareMetalHost with the latest ISO URL.
+
+The BMAC will also watch for the InfraEnv URL to change, and if there are any
```

Comment:
feels like it can delay boot time if the user adding nmstate configurations manually.
for example it will take 10 min to boot the machine then it may boot only after user finished to configure all his nmstate configurations.

Reply:
Yes, that's an inefficiency that this proposal accepts.